src/pyOpenMS/pyopenms/dataframes.py

```diff
@@ -304,6 +304,61 @@ def get_df(self, long : bool = False):
 
         return pd.DataFrame(data=((spec.getRT(), *spec.get_peaks()) for spec in self), columns=cols)
 
+    def get_massql_df(self):
+        """Exports data from MSExperiment to pandas DataFrames to be used with MassQL.
+
+        The Python module massql allows queries in mass spectrometry data (MS1 and MS2
+        data frames) in a SQL like fashion (https://github.com/mwang87/MassQueryLanguage).
+        
+        Returns:
+        ms1_df (pandas.DataFrame): peak data of MS1 spectra
+        ms2_df (pandas.DataFrame): peak data of MS2 spectra with precursor information
+        """
+        self.updateRanges()
+
+        def get_polarity(spectrum):
+            polarity = spectrum.getInstrumentSettings().getPolarity()
+            if polarity == IonSource.Polarity.POLNULL:
+                return 0
+            elif polarity == IonSource.Polarity.POSITIVE:
+                return 1
+            elif polarity == IonSource.Polarity.NEGATIVE:
+                return 2
+
+        def get_peak_arrays_from_spec(spec, scan_num):
+            if spec.getMSLevel() == 2:
+                ms2_data = (spec.getPrecursors()[0].getMZ(), self.getPrecursorSpectrum(scan_num)+1, spec.getPrecursors()[0].getCharge())
+            else:
+                ms2_data = ()
+            for i, peak in enumerate(spec):
+                peak_data = (peak.getIntensity(), peak.getMZ(), scan_num+1, spec.getRT()/60, get_polarity(spec))
+                yield peak_data + ms2_data
+
+        def get_spec_arr(spec, scan_num):
+            arr = np.asarray([peak_data for peak_data in get_peak_arrays_from_spec(spec, scan_num)], dtype='f')
+            arr = np.insert(arr, 1, arr[:,0]/np.amax(arr[:,0]), axis=1) # i_norm
+            arr = np.insert(arr, 2, arr[:,0]/np.sum(arr[:,0]), axis=1) # tic_i_norm
```

Comment:
this can be made better readable by making a pd.DataFrame earlier on.
you can then use e.g.
`arr['i_norm'] = arr[: , 0]/...`
to add new columns and use pd.concat to stack them later on

Reply:
Using pandas at this point was significantly slower then numpy, maybe you can check the code I posted below if that is what you had in mind?